code/week7/main.cpp

```diff
@@ -0,0 +1,206 @@
+#include <iostream>
+#include <set>
+#include <unordered_set>
+#include <map>
+#include <unordered_map>
+#include <list>
+#include <stack>
+#include <queue>
+#include <memory>
+
+#include "generic_class.h"
+
+// TODO clean this method up
+template<class T>
+void print_data(T iterator_begin, T iterator_end, std::string name) {
+    std::cout << "Printing " << name << std::endl;
+    for(;iterator_begin != iterator_end; iterator_begin++) {
+        std::cout << *iterator_begin << ", ";
+    }
+    std::cout << "\n\n";
+}
+/* with set<int> it would convert to something like this
+void print_data(set<int>::iterator iterator_begin, set<int>::iterator iterator_end, std::string name) {
+    std::cout << "Printing " << name << std::endl;
+    for(;iterator_begin != iterator_end; iterator_begin++) {
+        std::cout << *iterator_begin << ", ";
+    }
+    std::cout << "\n\n";
+}
+ */
+
+template <class Foo>
+Foo getMax(Foo a, Foo b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+/* when getMax<int> the method would look like
+int getMax(int a, int b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+/*
+when getMax<double> the method would look like
+double getMax(double a, double b) {
+    if(a < b) {
+        return b;
+    } else {
+        return a;
+    }
+}
+*/
+
+// you can have multiple templates in a function
+template <class Foo, class Bar>
+Bar getMin(Foo a, Bar b) {
```

Comment:
I think this just comes down to c++ letting you shoot yourself in the foot if you want to.